pyxem/utils/indexation_utils.py

```diff
@@ -923,3 +923,527 @@ def peaks_from_best_vector_match(single_match_result, library, rank=0):
+    r, theta = get_template_polar_coordinates(simulation, in_plane_angle=0.,
+                                              delta_r=delta_r, delta_theta=delta_theta)
+    r = np.rint(r).astype(np.int64)
+    theta = np.rint(theta).astype(np.int64)
+    condition = (r>0) & (r<polar_image.shape[1])
+    intensity = simulation.intensities
+    r = r[condition]
+    theta = theta[condition]
+    intensity = intensity[condition]
+    # norm of polar image
+    polar_image_norm = np.sqrt(np.sum(np.power(polar_image.ravel(), 2)))
+    polar_image = polar_image/polar_image_norm
+    correlation_array = _match_polar_to_polar_template(polar_image, r, theta, intensity)
+    angle_array = np.arange(correlation_array.shape[0])*delta_theta
+    return angle_array, correlation_array
+
+
+@njit(nogil=True)
+def _match_polar_to_polar_library(polar_image, r_templates, theta_templates, intensities_templates):
+    """Correlates a polar pattern to all polar templates and returns the
+    correlation index and best fitting in-plane angles for each template.
+    Each is returned as a 1D array"""
+    correlations = np.zeros(intensities_templates.shape[0])
+    angles = np.zeros(intensities_templates.shape[0])
+    d_theta = 360 / polar_image.shape[0]
+    for i in prange(intensities_templates.shape[0]):
+        intensities_template = intensities_templates[i]
+        r_template = r_templates[i]
+        theta_template = theta_templates[i]
+        match = _match_polar_to_polar_template(polar_image, r_template, theta_template, intensities_template)
+        correlations[i] = np.max(match)
+        angles[i] = np.argmax(match) * d_theta
+    return correlations, angles
+
+
+@njit
+def _get_correlation_at_angle(polar_image, r_templates, theta_templates, intensities, angle_shifts):
+    """Get the correlation between a polar image and polar templates at particular theta-shifts"""
+    correlations = np.zeros(r_templates.shape[0], dtype=np.float64)
+    for i in range(r_templates.shape[0]):
+        angle = angle_shifts[i]
+        r = r_templates[i]
+        theta = np.mod(theta_templates[i] + angle, polar_image.shape[0])
+        r = r.astype(np.int64)
+        theta = theta.astype(np.int64)
+        intensity = intensities[i]
+        image_intensities = _extract_pixel_intensities(polar_image, r, theta)
+        correlations[i] = np.sum(np.multiply(image_intensities, intensity))
+    return correlations
+    
+    
+def correlate_library_to_pattern(image, simulations, delta_r=1, delta_theta=1, find_direct_beam=True, **kwargs):
+    """
+    Get the best angle and associated correlation values, as well as the correlation with the inverted templates
+    
+    Parameters
+    ----------
+    image : 2d numpy.ndarray
+        The pattern
+    simulations : list of diffsims.sims.diffraction_simulation.DiffractionSimulation
+        The diffraction pattern simulation
+    delta_r : float
+        The sampling of the radial coordinate
```

Comment:
units?

Reply:
I don't know how to explain it best, maybe you can offer a suggestion. When converting to polar coordinates, an image of size (Y,X) gets converted to size (360/delta_theta, sqrt(X^2 + Y^2)/delta_r). So theta is trivially in coordinates of degrees, but r is in units of pixels of original image. Seems like quite a confusing mouthful.

Reply:
Hmm, might have to have a think about this.

Reply:
I think "pixels" is good enough here.